src/main/java/graphql/annotations/GraphQLAnnotations.java

```diff
@@ -638,27 +693,49 @@ protected static GraphQLFieldDefinition field(Method method) throws Instantiatio
+        String typeName = getTypeName(object);
+        if (typeRegistry.containsKey(typeName)) {
+            return (GraphQLInputObjectType) typeRegistry.get(typeName);
+        } else {
+            graphql.schema.GraphQLType graphQLType = getObject(object);
+            GraphQLInputObjectType inputObject = (GraphQLInputObjectType) getInputObject(graphQLType, "");
+            typeRegistry.put(inputObject.getName(), inputObject);
+            return inputObject;
+        }
+    }
+
     @Override
-    public GraphQLInputObjectType getInputObject(GraphQLObjectType graphQLType, String newNamePrefix) {
-        GraphQLObjectType object = graphQLType;
-        return new GraphQLInputObjectType("" + newNamePrefix + object.getName(), object.getDescription(),
-                object.getFieldDefinitions().stream().
-                        map(field -> {
-                            GraphQLOutputType type = field.getType();
-                            GraphQLInputType inputType;
-                            if (type instanceof GraphQLObjectType) {
-                                inputType = getInputObject((GraphQLObjectType) type, newNamePrefix);
-                            } else {
-                                inputType = (GraphQLInputType) type;
-                            }
-
-                            return new GraphQLInputObjectField(field.getName(), field.getDescription(), inputType, null);
-                        }).
-                        collect(Collectors.toList()));
+    public GraphQLInputType getInputObject(graphql.schema.GraphQLType graphQLType, String newNamePrefix) {
+        if (graphQLType instanceof GraphQLObjectType) {
+            GraphQLObjectType object = (GraphQLObjectType) graphQLType;
+            if (typeRegistry.containsKey(newNamePrefix + object.getName()) && typeRegistry.get(newNamePrefix + object.getName()) instanceof GraphQLInputType) {
+                return (GraphQLInputType) typeRegistry.get(newNamePrefix + object.getName());
+            }
+            GraphQLInputObjectType inputObjectType = new GraphQLInputObjectType(newNamePrefix + object.getName(), object.getDescription(),
+                    object.getFieldDefinitions().stream().
+                            map(field -> {
+                                GraphQLOutputType type = field.getType();
+                                GraphQLInputType inputType = getInputObject(type, newNamePrefix);
+                                return new GraphQLInputObjectField(field.getName(), field.getDescription(), inputType, null);
+                            }).
+                            collect(Collectors.toList()));
+            typeRegistry.put(inputObjectType.getName(), inputObjectType);
+            return inputObjectType;
+        } else if (graphQLType instanceof GraphQLList) {
+            return new GraphQLList(getInputObject(((GraphQLList)graphQLType).getWrappedType(), newNamePrefix));
```

Comment:
It looks like you dont handle recursive lists, like List<List<InputType>>

Reply:
the getInputObject is recursive, so List of List should be handled. I've added a unit test for that and it works. Do you have a code sample that does not work as expected ?